core/test/lib/libledger-test/ProxyHttpClient.cpp

```diff
@@ -1,28 +1,254 @@
+                    virtual unordered_map<string, string> getHeaders() {
+                        return _request->getHeaders();
+                    }
+
+                    virtual vector<uint8_t> getBody() {
+                        return _request->getBody();
+                    }
+
+                    virtual string getUrl() {
+                        return _request->getUrl();
+                    }
+
+                    virtual void complete(const shared_ptr<api::HttpUrlConnection> & response, const experimental::optional<api::Error> & error) {
+                        const string response_body = readResponseBody(response);
+                        _callback(to_string(getMethod()), getUrl(), getBody(), response_body);
+                        _request->complete(FakeUrlConnection::fromString(response_body), error);
+                    }
+
+                protected:
+                    static string readResponseBody(const shared_ptr<api::HttpUrlConnection> & response) {
+                        string response_body;
+                        while (true) {
+                            api::HttpReadBodyResult body = response->readBody();
+
+                            if (body.error)
+                                throw runtime_error(body.error->message);
+
+                            if (body.data) {
+                                const vector<uint8_t>& data = *(body.data);
+                                if (data.empty())
+                                    break;
+                                response_body += string((char*)data.data(), data.size());
+                                continue;
+                            }
+                            break;
+                        }
+                        return response_body;
+                    }
+
+                    const shared_ptr<api::HttpRequest> _request;
+                    callback_t _callback;
+                };
+            }
+
+            ProxyHttpClient::ProxyHttpClient(shared_ptr<api::HttpClient> httpClient)
                 : _httpClient(httpClient)
-            {}
+            {
+#ifdef UPDATE_HTTP_CACHE
+                _logger = make_shared<impl::TrafficLogger>();
+#endif
 
-            void ProxyHttpClient::execute(const std::shared_ptr<api::HttpRequest>& request) {
-                auto it = _cache.find(request->getUrl());
-                if (it != _cache.end()) {
-                    std::cout << "get response from cache : " << request->getUrl() << std::endl;
-                    request->complete(it->second, std::experimental::nullopt);
+#ifdef LOAD_HTTP_CACHE
+                if (!_logger) {
+                    loadCache(impl::TrafficLogger::getCacheFileName());
+                }
+#endif
+            }
+
+            void ProxyHttpClient::execute(const shared_ptr<api::HttpRequest>& request) {
+                auto vector_uint8_to_string = [](const vector<uint8_t>& v) {
+                    return string((const char*)v.data(), v.size());

```

Comment:
I prefer the following syntax (more aligned with STL and C++17/20 paradigms): `std::string(v.begin(), v.end())`